Error while exporting course with too long filename

noraiz-anwar opened the pull request:
## [TNL-6197](https://openedx.atlassian.net/browse/TNL-6197)

### Description

Generation of video name is restricted to less than 255 characters to avoid 'File name too long' error, as video name is eventually used to create transcript file while exporting course.
Firstly, transcript object is created with video name and at the time of export same name is used as a filename.

### How to Test?

**Stage** 
Following are the steps to follow in order to observe this behavior.
- In any course, create a video unit.
- Set any additional URL resource for this video, say www.abc.com/{give-any-name-more-than-255-chars-here}.mp4
- Upload a transcript for this video.
- Try to export this course.
You will not be able to export.

Note: One can use text from attached file to use in 'additional URL resource' to avoid typing.
[video_name.txt](https://github.com/edx/edx-platform/files/674120/video_name.txt)

  

**Sandbox**
Repeat above steps and observe course is successfully exported.
- [Studio](https://studio-noraiz-anwar.sandbox.edx.org)

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Style and readability review: @Ayub-Khan 
- [x] Code review: @attiyaIshaque 
- [x] Code review: @Qubad786

FYI: Tag anyone who might be interested in this PR here.

### Post-review
- [ ] Rebase and squash commits

attiyaIshaque commented:
@noraiz-anwar we would probably need simple unit tests  to check the restriction of characters.

noraiz-anwar commented:
jenkins run bokchoy

noraiz-anwar commented:
jenkins run python

awaisdar001 commented on cms/static/js/views/video/transcripts/utils.js:
```diff
@@ -129,7 +130,10 @@ define(['jquery', 'underscore', 'jquery.ajaxQueue'], function($) {
+                        /* avoid too long video name, as it will be used as filename for video's transcript
+                        and a filename can not be more that 255 chars, limiting here to 100.
+                        */
+                        video: match[1].slice(0, maxLengthVideoName),
```
this needs a method which can `slice` the matching string to max_length. We are slicing at two places, so it indicates to me that we can move the logic to a single place so that any change in required in future is done in that method.

noraiz-anwar replied:
Any future change can be placed in `maxLengthVideoName`. My point of view is, it will be an overkill to create a function to just call another function(`slice`). If you found it better to have a function anyway, I am open to create it.
Thanks

awaisdar001 commented on common/lib/xmodule/xmodule/video_module/transcripts_utils.py:
```diff
@@ -296,9 +296,11 @@ def copy_or_rename_transcript(new_name, old_name, item, delete_old=False, user=N
+    Slices each id by 100, restricting too long strings as video names.
     """
-    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0] for x in html5_sources]
+    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0][:100] for x in html5_sources]
```
Is this line going to fix the current issue? It seems the JS change is required for prevention and backend change is being made for the fix?

noraiz-anwar replied:
Currently there are two places where video name is extracted, one from JS at front end for operations like check/upload transcripts. One is here when someone hits save in editor for example after adding a new URL resource(also after uploading transcript). This new URL resource is supposed to have a transcript instance, so this method is used for extracting name for that transcript.

awaisdar001 commented on cms/static/js/spec/video/transcripts/utils_spec.js:
```diff
@@ -215,6 +215,39 @@ function($, _, Utils, _str) {
+                            '%s'
+                        ];
+                        return $.map(links, function(link) {
+                            return _str.sprintf(link, videoName);
```
`_str.sprintf(link, videoName);` what this line is doing? concatenating strings?

noraiz-anwar replied:
No, it is formatting the links with provided videoName

noraiz-anwar replied:
@attiyaIshaque I will rather say it formatting.

awaisdar001 commented on cms/static/js/spec/video/transcripts/utils_spec.js:
```diff
@@ -215,6 +215,39 @@ function($, _, Utils, _str) {
+                    it(link, function() {
+                        var result = Utils.parseHTML5Link(link);
+
+                        expect(result.video.length).toBeLessThan(255);
```
As we are allowing 100 as max file name size, shouldn't we also check for 100 here?

noraiz-anwar replied:
We can do this as you suggested. Reason for writing this is if some one wants to change length lets say within range of 255. We will not have to change this.

attiyaIshaque commented on cms/static/js/views/video/transcripts/utils.js:
```diff
@@ -129,7 +130,10 @@ define(['jquery', 'underscore', 'jquery.ajaxQueue'], function($) {
                 if (match) {
                     cache[url] = {
-                        video: match[1],
+                        /* avoid too long video name, as it will be used as filename for video's transcript
```
I didn't get this doc string. Please try to reword it. :)

noraiz-anwar commented on cms/djangoapps/contentstore/tests/test_transcripts_utils.py:
```diff
@@ -184,6 +186,14 @@ def clear_subs_content(self, youtube_subs):
+    def generate_random_string(self, string_length=100):
+        """
+        Generate random string
+        :param string_length: length of string to be generated, default 100
```
sure

awaisdar001 commented on cms/djangoapps/contentstore/tests/test_transcripts_utils.py:
```diff
@@ -184,6 +186,17 @@ def clear_subs_content(self, youtube_subs):
         for subs_id in youtube_subs.values():
             self.clear_sub_content(subs_id)
 
+    def generate_random_string(self, string_length=100):
```
What a valid reason one can give to this `string_length` variable having a default value of `100`?

noraiz-anwar replied:
Just a choice, no reason to have it 100. If you find any other appropriate value, please guide.

attiyaIshaque commented:
Looks good to me 👍

Qubad786 commented on cms/static/js/views/video/transcripts/utils.js:
```diff
@@ -110,6 +110,7 @@ define(['jquery', 'underscore', 'jquery.ajaxQueue'], function($) {
      */
         var _videoLinkParser = (function() {
             var cache = {};
+            var maxLengthVideoName = 150;
```
`maxVideoNameLength` seems more readable to me.

noraiz-anwar replied:
Yeah this is more readable. Going to change it.

noraiz-anwar commented:
jenkins run js

awaisdar001 commented on cms/djangoapps/contentstore/tests/test_transcripts_utils.py:
```diff
@@ -230,6 +243,17 @@ def test_subs_for_html5_vid_with_periods(self):
+        Filename can't be longer than 255 characters.
+        150 chars is agreed length.
+        """
+        html5_ids = transcripts_utils.get_html5_ids([self.generate_random_string(255)])
```
Can you use  `get_random_string` from `django.utils.crypto import`?

noraiz-anwar replied:
It is good to know about such method. Going to change it. Thanks

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run quality

awaisdar001 commented:
jenkins run quality

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run quality

noraiz-anwar commented:
jenkins run quality

adampalay commented on common/lib/xmodule/xmodule/video_module/transcripts_utils.py:
```diff
@@ -296,9 +296,11 @@ def copy_or_rename_transcript(new_name, old_name, item, delete_old=False, user=N
+    Slices each id by 150, restricting too long strings as video names.
     """
-    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0] for x in html5_sources]
+    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0][:150] for x in html5_sources]
```
@noraiz-anwar , how do we ensure that these are all unique?

noraiz-anwar replied:
`html5_sources` are basically other html5 URL resources provided for video by course team. Here at this line of code, we are extracting names(html5_ids) for videos from URLs. Why do you think we need to ensure about uniqueness of these html5_ids?

noraiz-anwar commented:
Actually we don't need uniqueness here,it is just parsing of whatever html
resources are passed as video links

On Jan 5, 2017 7:24 PM, "Adam" <notifications@github.com> wrote:

*@adampalay* commented on this pull request.
------------------------------

In common/lib/xmodule/xmodule/video_module/transcripts_utils.py
<https://github.com/edx/edx-platform/pull/14216#pullrequestreview-15311752>:

>      """
-    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0] for x in html5_sources]
+    html5_ids = [x.split('/')[-1].rsplit('.', 1)[0][:150] for x in
html5_sources]

@noraiz-anwar <https://github.com/noraiz-anwar> , how do we ensure that
these are all unique?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/edx/edx-platform/pull/14216#pullrequestreview-15311752>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AVT9VANoYEgHdGp5fnX3F9Vv85F8zHlpks5rPPz9gaJpZM4LWM5E>
.

cpennington commented:
This PR has been deployed to edx.org Staging in preparation for a release to production on Fri, Jan 6, 2017. If you feel you need to validate it manually, please do so by then, and contact me if there is an urgent enough issue to scrap the deployment.

cpennington commented:
This PR has been deployed to edx.org.

Ayub-Khan commented:
@cpennington we have a created another pr to revert this https://github.com/edx/edx-platform/pull/14275

cpennington commented:
@Ayub-Khan: Did this break production?

adampalay commented:
@cpennington , no, we just decided it wasn't the right fix.

cpennington commented:
Oh, ok.

noraiz-anwar commented:
@adampalay @cpennington revert PR is failing at bokchoy tests unexpectedly. I will be able merge it as soon as tests are passed.
